sbtc/src/leb128.rs

```diff
@@ -0,0 +1,556 @@
+    #[test_case(127, &[127], 1 ; "small value")]
+    #[test_case(0xFFFFFFFFFFFFFFFF, &[255, 255, 255, 255, 255, 255, 255, 255, 255, 1], 10 ; "large value")]
+    fn test_leb128_encode(value: u64, expected_bytes: &[u8], expected_len: usize) {
+        let mut bytes = Vec::new();
+        Leb128::encode_into(value, &mut bytes);
+        assert_eq!(bytes.as_slice(), expected_bytes);
+        assert_eq!(bytes.len(), expected_len);
+    }
+
+    /// Tests decoding of specific byte patterns
+    ///
+    /// Verifies correct value extraction and byte consumption tracking.
+    /// Includes error handling for empty inputs.
+    #[test_case(&[172, 2] => Ok((300, 2)) ; "medium value")]
+    #[test_case(&[127] => Ok((127, 1)) ; "small value")]
+    #[test_case(&[255, 255, 255, 255, 255, 255, 255, 255, 255, 1] => Ok((0xFFFFFFFFFFFFFFFF, 10)) ; "large value")]
+    #[test_case(&[] => Err(Error::EmptyInput) ; "empty input")]
+    #[test_case(&[0x7F, 0x00] => Ok((127, 1)) ; "value with trailing bytes")]
+    fn test_leb128_decode(bytes: &[u8]) -> Result<(u64, usize), Error> {
+        Leb128::try_decode(bytes)
+    }
+
+    /// Tests boundary conditions and unique encoding patterns
+    ///
+    /// Verifies important edge cases that property tests might miss:
+    /// - Multi-byte boundary values
+    /// - Powers of two near byte transitions
+    /// - Unusual bit patterns
+    #[test_case(&[0x80, 0x80, 0x01] => Ok((16384, 3)) ; "three byte value")]
+    #[test_case(&[0xFF, 0xFF, 0x03] => Ok((65535, 3)) ; "max two bytes as three")]
+    #[test_case(&[0x80, 0x80, 0x80, 0x80, 0x10] => Ok((1 << 32, 5)) ; "five byte power of two")] // 4294967296
+    #[test_case(&[0x80, 0x80, 0x80, 0x80, 0x20] => Ok((1 << 33, 5)) ; "five byte power of two plus one")] // 8589934592
+    fn test_leb128_decode_edge_cases(bytes: &[u8]) -> Result<(u64, usize), Error> {
+        Leb128::try_decode(bytes)
+    }
+
+    /// Tests error handling for malformed inputs
+    ///
+    /// Ensures robustness against truncated byte sequences and proper error
+    /// response for invalid inputs.
+    #[test_case(&[0x80] => Err(Error::IncompleteSequence); "incomplete byte sequence")]
+    #[test_case(&[0x80, 0x80] => Err(Error::IncompleteSequence); "truncated multi-byte")]
+    #[test_case(&[0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0xFF] => Err(Error::InvalidContinuation) ; "too many continuation bytes")]
+    #[test_case(&[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x01] => Err(Error::InvalidContinuation) ; "exceeds u64 (11 bytes)")]
+    #[test_case(&[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x01] => Err(Error::InvalidContinuation) ; "exceeds u64 (12 bytes)")]
+    fn test_leb128_decode_invalid(bytes: &[u8]) -> Result<(u64, usize), Error> {
+        Leb128::try_decode(bytes)
+    }
+
+    /// Tests position advancement in cursor reading
+    #[test]
+    fn test_cursor_position_tracking() {
```

Comment:
nit: can we add also a testcase for this where we have some garbage after the three valid ints? eg a continuation byte an nothing else

Reply:
Yup, added in the lower part of [993731d](https://github.com/stacks-network/sbtc/pull/1523/commits/993731d5d51c979c800378a2b1fbddc90211757f) as well